// aws-cdk/lib/api/cxapp/environments.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cxapi = require("@aws-cdk/cx-api");
const minimatch = require("minimatch");
async function globEnvironmentsFromStacks(appStacks, environmentGlobs, sdk) {
    if (environmentGlobs.length === 0) {
        environmentGlobs = ['**']; // default to ALL
    }
    const stacks = await appStacks.listStacks();
    const availableEnvironments = new Array();
    for (const stack of stacks) {
        const actual = await parseEnvironment(sdk, stack.environment);
        availableEnvironments.push(actual);
    }
    const environments = distinct(availableEnvironments).filter(env => environmentGlobs.find(glob => minimatch(env.name, glob)));
    if (environments.length === 0) {
        const globs = JSON.stringify(environmentGlobs);
        const envList = availableEnvironments.length > 0 ? availableEnvironments.map(env => env.name).join(', ') : '<none>';
        throw new Error(`No environments were found when selecting across ${globs} (available: ${envList})`);
    }
    return environments;
}
exports.globEnvironmentsFromStacks = globEnvironmentsFromStacks;
async function parseEnvironment(sdk, env) {
    const account = env.account === cxapi.UNKNOWN_ACCOUNT ? await sdk.defaultAccount() : env.account;
    const region = env.region === cxapi.UNKNOWN_REGION ? await sdk.defaultRegion() : env.region;
    if (!account || !region) {
        throw new Error(`Unable to determine default account and/or region`);
    }
    return {
        account, region,
        name: cxapi.EnvironmentUtils.format(account, region)
    };
}
/**
 * Given a set of "<account>/<region>" strings, construct environments for them
 */
function environmentsFromDescriptors(envSpecs) {
    if (envSpecs.length === 0) {
        throw new Error(`Either specify an app with '--app', or specify an environment name like '123456789012/us-east-1'`);
    }
    const ret = new Array();
    for (const spec of envSpecs) {
        const parts = spec.split('/');
        if (parts.length !== 2) {
            throw new Error(`Expected environment name in format '<account>/<region>', got: ${spec}`);
        }
        ret.push({
            name: spec,
            account: parts[0],
            region: parts[1]
        });
    }
    return ret;
}
exports.environmentsFromDescriptors = environmentsFromDescriptors;
/**
 * De-duplicates a list of environments, such that a given account and region is only represented exactly once
 * in the result.
 *
 * @param envs the possibly full-of-duplicates list of environments.
 *
 * @return a de-duplicated list of environments.
 */
function distinct(envs) {
    const unique = {};
    for (const env of envs) {
        const id = `${env.account || 'default'}/${env.region || 'default'}`;
        if (id in unique) {
            continue;
        }
        unique[id] = env;
    }
    return Object.values(unique);
}
